utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/JsonSchemaOutputValidator.scala

```diff
@@ -72,53 +68,133 @@ class JsonSchemaOutputValidator(validationMode: ValidationMode) extends LazyLogg
+    schemasToCheck.exists(schema => validateTypingResult(typingResult, schema, None).isValid)
+  }
+
+  private def validateRecordInputType(typingResult: TypedObjectTypingResult, schema: ObjectSchema, path: Option[String]): ValidatedNel[OutputValidatorError, Unit] = {
+    val explicitProps = schema.getPropertySchemas.asScala.toMap
+    val requiredProps = schema.getRequiredProperties.asScala.toSet
+    val schemaFieldsValidation = validateRecordExplicitProperties(typingResult, explicitProps, path)
+
+    val requiredPropsV = validateRecordRequiredProps(typingResult, explicitProps, requiredProps, path)
+    val redundantPropsV = validateRecordRedundantProps(typingResult, schema, explicitProps, path)
+    val (patternPropsV, inputFieldsMatchedByPatternProps) = validateRecordPatternProps(typingResult, schema, path)
+    val foundAdditionalProps = findRecordAdditionalProps(typingResult, explicitProps.keySet, inputFieldsMatchedByPatternProps)
+    val additionalPropsV = validateRecordAdditionalProps(schema, path, foundAdditionalProps)
+
+    requiredPropsV.combine(schemaFieldsValidation)
+      .combine(redundantPropsV)
+      .combine(patternPropsV)
+      .combine(additionalPropsV)
+  }
+
+  private def validateRecordRequiredProps(typingResult: TypedObjectTypingResult, explicitPropsSchemas: Map[String, Schema], requiredProps: Set[String], path: Option[String]): ValidatedNel[OutputValidatorMissingFieldsError, Unit] = {
+    val requiredPropsNames = if (validationMode == ValidationMode.strict) {
+      explicitPropsSchemas.keys.toSet
+    } else {
+      requiredProps
     }
+    val missingProps = requiredPropsNames.filterNot(typingResult.fields.contains)
+    condNel(missingProps.isEmpty, (), OutputValidatorMissingFieldsError(buildFieldsPaths(missingProps, path)))
+  }
 
-    requiredFieldsValidation combine schemaFieldsValidation combine redundantFieldsValidation combine additionalFieldsValidation
+  private def validateRecordExplicitProperties(typingResult: TypedObjectTypingResult, schemaFields: Map[String, Schema], path: Option[String]): ValidatedNel[OutputValidatorError, Unit] = {
+    validateFieldsType(typingResult.fields.filterKeys(schemaFields.contains), schemaFields, path)
+  }
+
+  private def validateRecordRedundantProps(typingResult: TypedObjectTypingResult, schema: ObjectSchema, schemaFields: Map[String, Schema], path: Option[String]): ValidatedNel[OutputValidatorRedundantFieldsError, Unit] = {
+    val redundantFields = typingResult.fields.keySet.diff(schemaFields.keySet)
+    condNel(redundantFields.isEmpty || schema.permitsAdditionalProperties(), (), OutputValidatorRedundantFieldsError(buildFieldsPaths(redundantFields, path)))
+  }
+
+  private def validateRecordPatternProps(typingResult: TypedObjectTypingResult, schema: ObjectSchema, path: Option[String]): (Validated[NonEmptyList[OutputValidatorError], Unit], Set[String]) = {
+    val fieldsWithMatchedPatternsProperties = typingResult.fields.toList
```

Comment:
I think it would be a bit easier to operate on map here?

Reply:
as we agreed, I'll ignore this one 😄